services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt

```diff
@@ -113,38 +167,287 @@ class HakemusServiceITest : DatabaseTest() {
+                            .copy(areas = listOf(intersectingArea)),
+                        USERID
+                    )
+                }
+
+            Assertions.assertEquals(
+                """Invalid geometry received when updating application id=${application.id},
+                | reason = Self-intersection,
+                | location = {"type":"Point","coordinates":[25494009.65639264,6679886.142116806]}"""
+                    .trimMargin()
+                    .replace("\n", ""),
+                exception.message
+            )
+        }
+
+        @Test
+        fun `throws exception when the request has a persisted contact but the application does not`() {
+            val application = createHakemus()
+            val requestYhteystietoId = UUID.randomUUID()
+
+            val exception =
+                assertThrows<InvalidHakemusyhteystietoException> {
+                    hakemusService.updateHakemus(
+                        application.id!!,
+                        HakemusUpdateRequestFactory
+                            .createJohtoselvityshakemusUpdateRequestFromApplicationEntity(
+                                application
+                            )
+                            .copy(
+                                customerWithContacts =
+                                    HakemusUpdateRequestFactory.createCustomerWithContactsRequest(
+                                        yhteystietoId = requestYhteystietoId
+                                    )
+                            ),
+                        USERID
+                    )
+                }
+
+            assertThat(exception.message)
+                .isEqualTo(
+                    "Invalid hakemusyhteystieto received when updating application id=${application.id}, role=HAKIJA, yhteystietoId=null, newId=$requestYhteystietoId"
+                )
+        }
+
+        @Test
+        fun `throws exception when the request has different persisted contact than the application`() {
+            val application = createHakemusWithHakija()
+            val originalYhteystietoId = hakemusyhteystietoRepository.findAll().first().id
+            val requestYhteystietoId = UUID.randomUUID()
+
+            val exception =
+                assertThrows<InvalidHakemusyhteystietoException> {
+                    hakemusService.updateHakemus(
+                        application.id!!,
+                        HakemusUpdateRequestFactory
+                            .createJohtoselvityshakemusUpdateRequestFromApplicationEntity(
+                                application
+                            )
+                            .copy(
+                                customerWithContacts =
+                                    HakemusUpdateRequestFactory.createCustomerWithContactsRequest(
+                                        yhteystietoId = requestYhteystietoId
+                                    )
+                            ),
```

Comment:
I think creating the request should be in the arrange part of the AAA pattern. Especially if creating it is complicated like this. The act part should be simple.

Reply:
You're right. Thanks!